### PR TITLE
Add DidChangeWatchedFilesRegistrationOptions

### LINKED ISSUE #229: Add DidChangeWatchedFilesRegistrationOptions
https://microsoft.github.io/language-server-protocol/specifications/specification-current/#workspace_didChangeWatchedFiles

It is either not implemented or I cant find it

### PR DESCRIPTION
**Pull request by fendor**
Makes it possible to dynamically register capabilities for watching file changes.

Closes #229

### CODE REVIEW

File: haskell-lsp-types/src/Language/Haskell/LSP/Types/DataTypesJSON.hs [hunk 1]
```diff
@@ -1093,6 +1093,100 @@ type UnregisterCapabilityRequest = RequestMessage ServerMethod UnregistrationPar
...
+      deleteNum = if _watchDelete wk then 4 else 0
+
+instance A.FromJSON WatchKind where
+  parseJSON (A.Number n)
+    | 1 <= n && n <= 7 =
+      let
+        (wkDelete, n1) = if 4 <= n then (True, n-4) else (False, n)
+        (wkChange, n2) = if 2 <= n1 then (True, n1-2) else (False, n1)
+        (wkCreate, _) = if 1 <= n2 then (True, n2-1) else (False, n2)
+      in
```
Comment (fendor) [inline haskell-lsp-types/src/Language/Haskell/LSP/Types/DataTypesJSON.hs]: Really ugly, but I could not think of a way without introducing loads of types and boilerplate.

Comment (bubba) [reply haskell-lsp-types/src/Language/Haskell/LSP/Types/DataTypesJSON.hs]: This looks like a bit mask, so a common way would be with a [bitwise and](https://hackage.haskell.org/package/base-4.12.0.0/docs/Data-Bits.html#v:.-38-.)

Something like

```haskell
parseJSON (A.Number n) = pure $ WatchKind (n .&. 1) (n .&. 2) (n .&. 4)
```

Comment (fendor) [reply haskell-lsp-types/src/Language/Haskell/LSP/Types/DataTypesJSON.hs]: Agreed, however, `Scientific` doesnt implement `Bits` :/ we could translate it to Int, but I think that is unsafe.

Comment (bubba) [reply haskell-lsp-types/src/Language/Haskell/LSP/Types/DataTypesJSON.hs]: Argh I forgot about Javascript numbers. Surely anything thats not an integer is invalid anyway right? And should fail parsing

Comment (alanz) [reply haskell-lsp-types/src/Language/Haskell/LSP/Types/DataTypesJSON.hs]: Given it should be an integer, you can probably use one of the conversion functions to integer, then bitwise it.

Comment (fendor) [reply haskell-lsp-types/src/Language/Haskell/LSP/Types/DataTypesJSON.hs]: `testBit` is the function to use!

File: haskell-lsp-types/src/Language/Haskell/LSP/Types/DataTypesJSON.hs [hunk 2]
```diff
@@ -1093,6 +1096,96 @@ type UnregisterCapabilityRequest = RequestMessage ServerMethod UnregistrationPar
...
+    where
+      createNum = if _watchCreate wk then 0x1 else 0x0
+      changeNum = if _watchChange wk then 0x2 else 0x0
+      deleteNum = if _watchDelete wk then 0x4 else 0x0
+
+instance A.FromJSON WatchKind where
+  parseJSON (A.Number n)
+    | Right i <- floatingOrInteger n :: Either Double Int
+    , 0 <= i && i <= 7 =
+        pure $ WatchKind (testBit i 0x0) (testBit i 0x1) (testBit i 0x2)
```
Comment (bubba) [inline haskell-lsp-types/src/Language/Haskell/LSP/Types/DataTypesJSON.hs]: Should this be 0x1, 0x2, 0x4 for the values in watch kind?

Comment (bubba) [reply haskell-lsp-types/src/Language/Haskell/LSP/Types/DataTypesJSON.hs]: Never mind, I see it’s the nth bit. LGTM

Comment (fendor) [reply haskell-lsp-types/src/Language/Haskell/LSP/Types/DataTypesJSON.hs]: It took me three attempts to get it right. That's why there is a jsonSpec test now :)

**Comment by bubba** [comment]
Is this good to merge?

**Comment by fendor** [comment]
Actually, did anyone cross check that somehow if I got it right? Where should I add unit-tests for it?

**Comment by bubba** [comment]
This is just adding a new data type, so I think the round trip unit test should suffice

**Pull request merged**